### DISCUSSION TITLE
Dynamically create content of pages

### DISCUSSION THREAD

**Discussion by faanskit** [Hieromon/AutoConnect#554] [How can I?] [answered]
Hi,

Great code! Love it! Using it for a home project but my C++ competence is weak/old/rusty. 20+ years since I worked with c++ and coding, and a lot have happened with the language.

I'm trying to dynamically populate create pages with elements, but my c++ competence is simply not good enough.

My current static variant works fine, i.e. [a variant of] this:
```
AutoConnectAux* MyPortal::configPage(){
    std::vector<String> providers = getProviderMenuStr();
    std::vector<String> areas = getMarketMenuStr();
    AutoConnectElement* myProv = new AutoConnectSelect("Provider", providers, "Select provider", 1);
    AutoConnectElement* myArea= new AutoConnectSelect("Area", areas, "Select area", 1);
    return new AutoConnectAux("/area", "Configuration...", true, {*myProv , *myArea});;
}
```

To make it more dynamic, I intended to build a vector of AutoConnectElement(s) and later throw that at AutoConnectAux. I've tried many variants, and I cannot simply make it work/compile.

I've tried to understand how to make this work with AutoConnectElementVT but cannot find any examples to lean against.

My trial-and-error has now reached an end and hopefully someone can step in and assist. Here's a sample code that will not compile.

```
    const AutoConnectElementVT selectElements = AutoConnectElementVT();
    for(auto menuHead : getMainMenuItems()){
            switch(static_cast<ScreenOption>(menuHead.option)){
                    case ScreenOption::marketArea:
                        selectElements.push_back(new AutoConnectSelect(menuHead.str.c_str(), getMarketMenuStr(), menuHead.str.c_str()));
                        break;
            }
    }
    return new AutoConnectAux("/area", "Configuration...", true, selectElements );;
```

This variant throws:
```
src/myportal.cpp:96:137: error: no matching function for call to 'std::vector<std::reference_wrapper<AutoConnectElementJson> >::push_back(AutoConnectSelect*) const'
```
Removing the `new` from the .push_back() throws this compiler error:
```
src/myportal.cpp:96:133: error: passing 'const AutoConnectElementVT' {aka 'const std::vector<std::reference_wrapper<AutoConnectElementJson> >'} as 'this' argument discards qualifiers 
[-fpermissive]
```
So, now I need to stop my _trial-and-error_ and ask for help.

Thanks in advance,
/Marcus

**Comment by Hieromon** [comment, marked as answer]
@faanskit, Thank you.
So, is it like this?

```cpp
#include <Arduino.h>
#include <vector>
#include <WiFi.h>
#include <WebServer.h>
#include <AutoConnect.h>

AutoConnect portal;

void getProviderMenuStr(std::vector<String>& providers) {
  providers.push_back("foo");
  providers.push_back("bar");
  providers.push_back("qux");
}

void getMarketMenuStr(std::vector<String>& areas) {
  areas.push_back("Johnson City");
  areas.push_back("East HIlls");
  areas.push_back("North Castle");
}

class MyPortal {
 public:
  MyPortal() {}
  ~MyPortal() {}
  AutoConnectAux* configPage(void);
};

AutoConnectAux* MyPortal::configPage(void) {
  std::vector<String> providers;
  std::vector<String> areas;
  getProviderMenuStr(providers);
  getMarketMenuStr(areas);
  AutoConnectSelect* myProv = new AutoConnectSelect("Provider", providers, "Select provider", 1);
  AutoConnectSelect* myArea = new AutoConnectSelect("Area", areas, "Select area", 1);
  return new AutoConnectAux("/area", "Config...", true, {*myProv, *myArea});
}

void setup() {
  delay(1000);
  Serial.begin(115200);
  Serial.println();

  MyPortal  myPortal;
  portal.join({ *myPortal.configPage() });

  portal.begin();
}

void loop() {
  portal.handleClient();
}
```

**Reply by faanskit** [reply]
Thanks @Hieromon  for reaching out so quickly. The problem is a different one, but I do use it like in your example as well.

In essence, I try to dynamically create the `AutoConnectSelect` items.

More like this: (**NOTE**: This code does not compile and is likely stupid :-) )

```c++
#include <Arduino.h>
#include <vector>
#include <WiFi.h>
#include <WebServer.h>
#include <AutoConnect.h>

AutoConnect portal;

void getAcItems(std::vector<AutoConnectSelect*>& acItems) {
  AutoConnectSelect* pItem = new AutoConnectSelect("Provider", {"P1", "P2"}, "Select provider", 1);
  acItems.push_back(pItem);
  AutoConnectSelect* aItem = new AutoConnectSelect("Area", {"A1", "A2"}, "Select area", 1);
  acItems.push_back(aItem);
}

class MyPortal {
 public:
  MyPortal() {}
  ~MyPortal() {}
  AutoConnectAux* configPage(void);
};

AutoConnectAux* MyPortal::configPage(void) {
  std::vector<AutoConnectSelect*> acItems;
  getAcItems(acItems);
  return new AutoConnectAux("/area", "Config...", true, acItems);
}

void setup() {
  delay(1000);
  Serial.begin(115200);
  Serial.println();

  MyPortal  myPortal;
  portal.join({ *myPortal.configPage() });

  portal.begin();
}

void loop() {
  portal.handleClient();
}
```

Complier throws this:
```
src/main.cpp: In member function 'AutoConnectAux* MyPortal::configPage()':
src/main.cpp:26:64: error: no matching function for call to 'AutoConnectAux::AutoConnectAux(const char [6], const char [10], bool, std::vector<AutoConnectSelectJson*>&)'
   return new AutoConnectAux("/area", "Config...", true, acItems);
```
Somehow I think `AutoConnectElementVT` should be use, but cannot figure it out.

/Marcus

**Reply by Hieromon** [reply]
@faanskit, If you put your own class into `std::vector`, you need either a copy constructor or a move constructor. But AutoConnectElement doesn't implement it to save code. So _AutoConnectAux_ has a `std::vector` wrapping the _AutoConnectElement_ (including variants) in a `std::reference_wrapper`. That's `AutoConnectElementVT`.

https://github.com/Hieromon/AutoConnect/blob/80594fa132c51aa77f13846c7311a10e9e114c7a/src/AutoConnectAux.h#L33

Maybe something like this?

```cpp
#include <Arduino.h>
#include <vector>
#include <WiFi.h>
#include <WebServer.h>
#include <AutoConnect.h>

AutoConnect portal;

AutoConnectElementVT  getAcItems(void) {
  AutoConnectElementVT  items;

  AutoConnectElement* pItem(new AutoConnectSelect("Provider", {"P1", "P2"}, "Select provider", 1));
  std::reference_wrapper<AutoConnectElement>  pElm = std::ref(*pItem);
  items.push_back(pElm);

  AutoConnectElement* aItem(new AutoConnectSelect("Area", {"A1", "A2"}, "Select area", 1));
  std::reference_wrapper<AutoConnectElement>  aElm = std::ref(*aItem);
  items.push_back(aElm);

  return items;
}

class MyPortal {
 public:
  MyPortal() {}
  ~MyPortal() {}
  AutoConnectAux* configPage(void);
};

AutoConnectAux* MyPortal::configPage(void) {
  AutoConnectAux* aux(new AutoConnectAux("/area", "Config...", true));
  aux->add(getAcItems());
  return aux;
}

void setup() {
  delay(1000);
  Serial.begin(115200);
  Serial.println();

  MyPortal  myPortal;
  portal.join(*myPortal.configPage());

  portal.begin();
}

void loop() {
  portal.handleClient();
}
```

**NOTE:** The code above is deprecated. Ideally, you should use smart pointers to allocate AutoConnectrElements dynamically.

**Reply by faanskit** [reply]
@Hieromon **thanks a million** for your time and effort. I shall try this later today, but from the looks of it I can surely use it for my purposes. I would never have figured that out without your support.

I also need to learn about smart pointers I understand. That did not exist back-in-the-days. From a quick google, it looks quite straight forward. But I guess there will a few pitfalls for me..

Again, **thanks**!

**Reply by faanskit** [reply]
@Hieromon,
I can confirm that the propsal worked on a ESP32/az-delivery-devkit-v4, and with some tweaks it match my needs.
Below is a variant of the code, still without smart pointers, that more reflects what my intentions were.

Had to do a small work-around using `{}` since I don't fully understand this. But, it now serves my needs...

Leaving this for whomever may need it...

**Thanks!**
/Marcus

```c++
#include <Arduino.h>
#include <vector>
#include <WiFi.h>
#include <WebServer.h>
#include <AutoConnect.h>
#define PROVIDER 0
#define AREA 1

AutoConnect portal;
struct MenuItem {
  String str;
  int option;
};

std::vector<MenuItem> menuItems = {{"Provider", PROVIDER}, {"Area", AREA}};

std::vector<String> getProviders(){
  std::vector<String> providers = {"Provider 1", "Provider 2"};
  return providers;
}

std::vector<String> getAreas(){
  std::vector<String> areas = {"Area 1", "Area 2", "Area 3"};
  return areas;
}

AutoConnectElementVT  getAcItems(void) {
  AutoConnectElementVT  items;
  AutoConnectElement* pItem;
  for(auto item : menuItems){
    switch(item.option){
      case PROVIDER:
      { // Added bracket only to remove these errors:
        //   error: jump to case label
        //   error: redeclaration of 'std::reference_wrapper<AutoConnectElementJson> pElm'
        log_i("%s: {%s, %s}", item.str.c_str(), getProviders()[0], getProviders()[1]);
        pItem = new AutoConnectSelect(item.str.c_str(), getProviders(), "Select provider", 1);
        std::reference_wrapper<AutoConnectElement>  pElm = std::ref(*pItem);
        items.push_back(pElm);      
        break;
      }
      case AREA:
      {
        log_i("%s: {%s, %s, %s}", item.str.c_str(), getAreas()[0], getAreas()[1], getAreas()[2]);
        pItem = new AutoConnectSelect(item.str.c_str(), getAreas(), "Select area", 1);
        std::reference_wrapper<AutoConnectElement>  pElm = std::ref(*pItem);
        items.push_back(pElm);      
        break;
      }
    }
  }
  return items;
}

class MyPortal {
 public:
  MyPortal() {}
  ~MyPortal() {}
  AutoConnectAux* configPage(void);
};

AutoConnectAux* MyPortal::configPage(void) {
  AutoConnectAux* aux(new AutoConnectAux("/area", "Config...", true));
  aux->add(getAcItems());
  return aux;
}

void setup() {
  delay(1000);
  Serial.begin(115200);
  Serial.setDebugOutput(true);
  Serial.println("Starting...");

  MyPortal  myPortal;
  portal.join({ *myPortal.configPage() });

  portal.begin();
}

void loop() {
  portal.handleClient();
}
```

```
[env:az-delivery-devkit-v4]
platform = espressif32
board = az-delivery-devkit-v4
framework = arduino
monitor_speed = 115200
upload_speed = 460800
lib_deps = hieromon/AutoConnect@^1.4.1
build_flags = 
	-DCORE_DEBUG_LEVEL=ARDUHAL_LOG_LEVEL_INFO
```

**Reply by Hieromon** [reply]
I'm happy to have been a small help. So marked as an answer.

**Comment by Hieromon** [comment]
@faanskit 
> I also need to learn about smart pointers I understand. That did not exist back-in-the-days. From a quick google, it looks quite straight forward. But I guess there will a few pitfalls for me..

It may not be necessary, but I don't think you should rely too much on what's on the web to understand modern C++. C++ has undergone a major transformation with C++11. It has evolved abstraction modeling backed by powerful typing. As a result, copying and moving objects is a deliberate choice of the programmer in modern C++ programming. The compiler reads the intent and (consequently) decides whether to apply the pointer operation or leave the hidden temporary instance on the stack. For example, the type `std::vector<String>` is not "an object of the vector set of String"; it is a single type, just `std::vector<Stirng>`. It is a "type" in itself that represents a vector data structure. 

Therefore, manipulating a pointer to an object of type AutoConnectElement with the intention of a collection set of data will not work.

In your case, your knowledge of 90's C++ (or rather, C) is so deep that you probably took the approach of fixing the instance to a certain location in memory space and then manipulating the data structure with a pointer. That is a correct insight, but in modern C++, there are cases where that is counterproductive. In particular, you need to be careful in cases where **STL** is used a lot.

It is also about the appropriateness of a function returning an object. Local variables of primitive types are allocated on the stack and are undefined after the function returns. In the case of named objects, however, the value can be referenced without copying cost if the appropriate code is written. The return type of `getAcItems` is AutoConnectElementVT, which is a named return value of type `std::vector<std::reference_wrapper>>`. This story is closely related to **Rvalue References** with copies and moves. It is no exaggeration to say that this is the essence of C++11.

If you are interested, I suggest you take a look at a book, **Effective Modern C++**.

Regards,
